File: vito/nativecode/src/main/java/com/facebook/fresco/vito/nativecode/CircularBitmapTransformation.java
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.fresco.vito.nativecode;

import android.graphics.Bitmap;
import com.facebook.imagepipeline.nativecode.NativeRoundingFilter;
import com.facebook.imagepipeline.transformation.BitmapTransformation;
import com.facebook.imagepipeline.transformation.CircularTransformation;
import com.facebook.infer.annotation.Nullsafe;
import javax.annotation.Nullable;

@Nullsafe(Nullsafe.Mode.LOCAL)
public class CircularBitmapTransformation implements BitmapTransformation, CircularTransformation {

  private final boolean mAntiAliased;

  private final boolean mUseFastNativeRounding;

  public CircularBitmapTransformation(boolean antiAliased) {
    this(antiAliased, false);
  }

  public CircularBitmapTransformation(boolean antiAliased, boolean useFastNativeRounding) {
    mAntiAliased = antiAliased;
    mUseFastNativeRounding = useFastNativeRounding;
  }

  @Override
  public void transform(Bitmap bitmap) {
    if (mUseFastNativeRounding) {
      NativeRoundingFilter.toCircleFast(bitmap, mAntiAliased);
    } else {
      NativeRoundingFilter.toCircle(bitmap, mAntiAliased);
    }
  }

  @Override
  public boolean modifiesTransparency() {
    return true;
  }

  @Override
  public boolean equals(@Nullable Object obj) {
    if (this == obj) {
      return true;
    }
    if (obj == null || getClass() != obj.getClass()) return false;
    CircularBitmapTransformation other = (CircularBitmapTransformation) obj;

    return mAntiAliased == other.mAntiAliased;
  }

  public boolean isAntiAliased() {
    return mAntiAliased;
  }

  @Override
  public int hashCode() {
    return mAntiAliased ? 1 : 0;
  }
}

File: vito/nativecode/src/main/java/com/facebook/fresco/vito/nativecode/CircularBitmapTransformation.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.fresco.vito.nativecode

import android.graphics.Bitmap
import com.facebook.imagepipeline.nativecode.NativeRoundingFilter
import com.facebook.imagepipeline.transformation.BitmapTransformation
import com.facebook.imagepipeline.transformation.CircularTransformation

@Suppress("KtDataClass")
data class CircularBitmapTransformation(
    val isAntiAliased: Boolean = true,
    private val useFastNativeRounding: Boolean = false
) : BitmapTransformation, CircularTransformation {

  override fun transform(bitmap: Bitmap) {
    if (useFastNativeRounding) {
      NativeRoundingFilter.toCircleFast(bitmap, isAntiAliased)
    } else {
      NativeRoundingFilter.toCircle(bitmap, isAntiAliased)
    }
  }

  override fun modifiesTransparency(): Boolean = true

  override fun equals(other: Any?): Boolean {
    if (this === other) {
      return true
    }
    if (javaClass != other?.javaClass) {
      return false
    }

    val otherTransformation: CircularBitmapTransformation = other as CircularBitmapTransformation

    return isAntiAliased == otherTransformation.isAntiAliased &&
        useFastNativeRounding == otherTransformation.useFastNativeRounding
  }

  override fun hashCode(): Int {
    var result = isAntiAliased.hashCode()
    result = 31 * result + useFastNativeRounding.hashCode()
    return result
  }
}
